### PR TITLE
ENT-3832: Thread-safe double ended queue

### PR DESCRIPTION
**Pull request by karlhto**
Currently WIP. Functionality is fine, but a lot of the code could be standardised a lot more.

Also needs tests.

@olehermanse or someone else, add WIP, please.

### CODE REVIEW

**Comment by olehermanse** [comment]
@karlhto I changed the title, please update commit also, it is an ENT ticket, not CFE.

**Comment by olehermanse** [comment]
This pull request **introduces 1 alert** when merging a02724962b529426422dc1c6d5bcb2d7010976ea into 414a4a633437f45a113d1eb6372d98076b5914f3 - [view on LGTM.com](https://lgtm.com/projects/g/cfengine/core/rev/pr-91b392d82d70c604dcb01af7e3f55e2a4c24da08)

**new alerts:**

* 1 for For loop variable changed in body

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by olehermanse** [comment]
This pull request **introduces 1 alert** when merging 2fffa2ffaba32d52588f8bbb9c766c2b4097110a into 414a4a633437f45a113d1eb6372d98076b5914f3 - [view on LGTM.com](https://lgtm.com/projects/g/cfengine/core/rev/pr-1c4ecbaa171b418f73df437173a19573267f53d2)

**new alerts:**

* 1 for For loop variable changed in body

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by karlhto** [comment]
@cf-bottom Jenkins with exotics, please.

**Comment by karlhto** [comment]
@olehermanse @vpodzime This is no longer WIP, and is ready for review.

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2077/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/2077/

File: tests/unit/Makefile.am [hunk 1]
```diff
@@ -90,7 +90,7 @@ libdb_la_SOURCES += ../../libpromises/cf3globals.c
 endif
 libdb_la_LIBADD = libstr.la ../../libutils/libutils.la
 #libdb_la_CPPFLAGS = $(LMDB_CPPFLAGS) $(TOKYOCABINET_CPPFLAGS) $(QDBM_CPPFLAGS)
-# Make sure that source files are compiled to separate object files 
+# Make sure that source files are compiled to separate object files
```
Comment (karlhto) [inline tests/unit/Makefile.am]: Uh. I have configured vim to remove spaces at the end of lines when saving. Should I exclude this from the PR?

**Comment by karlhto** [comment]
Jenkins error seems unrelated

File: tests/unit/Makefile.am (hunk 1 above)
Comment (vpodzime) [reply tests/unit/Makefile.am]: Ideally, put changes like this into a separate commit like "whitespace cleanup".

Comment (olehermanse) [reply tests/unit/Makefile.am]: "Always" use `git add -p`. Then you are concious about changes like this. You can either skip it, or make it a separate commit.

Comment (vpodzime) [reply tests/unit/Makefile.am]: Or `git commit -p`.

Comment (vpodzime) [reply tests/unit/Makefile.am]: But definitely not a big deal in this case with just one line changed.

**Comment by olehermanse** [comment]
@cf-bottom run in jenkins again with exotics.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2171/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/2171/

File: libutils/threaded_deque.c [hunk 2]
```diff
@@ -0,0 +1,641 @@
...
+                    return false;
+                }
+            } while (deque->size != 0);
+            // Reevaluate predicate to protect against spurious wakeups
+        }
+        else
+        {
+            ret = false;
+        }
+    }
```
Comment (vpodzime) [inline libutils/threaded_deque.c]: I think this would be nicer if it was restructured to deal with the "easy" cases first. So things like
```
if (deque->size == 0)
{
    ThreadUnlock(deque->lock);
    return true;
}
```
and similarly for `timeout`.

File: libutils/threaded_deque.c [hunk 3]
```diff
@@ -0,0 +1,641 @@
...
+        do
+        {
+            deque->ItemDestroy(deque->data[i++]);
+            i %= deque->capacity;
+        } while (i != end);
+    }
+}
+
+/**
+  @brief Either expands capacity of dedeque, or shifts tail to beginning.
```
Comment (vpodzime) [inline libutils/threaded_deque.c]: *dedeque*

File: libutils/threaded_deque.c [hunk 4]
```diff
@@ -0,0 +1,641 @@
...
+    {
+        if (deque->right <= deque->left)
+        {
+            size_t old_capacity = deque->capacity;
+
+            deque->capacity *= EXPAND_FACTOR;
+            deque->data = xrealloc(deque->data,
+                                   sizeof(void *) * deque->capacity);
+
+            memmove(deque->data + old_capacity, deque->data,
```
Comment (vpodzime) [inline libutils/threaded_deque.c]: Aren't the arguments here in a wrong order? Also, is the number of items to be moved really always given by `deque->right`?

File: libutils/threaded_deque.h [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+/**
+  @brief Destroys the deque and frees the memory it occupies.
+  @warning ThreadedDeque should only be destroyed if all threads are joined
+  @param [in] deque  The deque to destroy.
+  */
+void ThreadedDequeDestroy(ThreadedDeque *deque);
+
+/**
+  @brief Frees the memory allocated for the data pointer and the struct itself.
```
Comment (vpodzime) [inline libutils/threaded_deque.h]: *memory allocated for the data pointer*? *pointer***s** maybe?

File: tests/unit/threaded_deque_test.c [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+
+static void test_push_pop(void)
+{
+    // Initialised with DEFAULT_CAPACITY = 16
+    ThreadedDeque *deque = ThreadedDequeNew(0, free);
+    // |^---------------|
+    ThreadedDequePushLeft(deque, xstrdup("1"));
+    // |<-------------->|
+    ThreadedDequePushRight(deque, xstrdup("2"));
+    // |x<------------->|
```
Comment (vpodzime) [inline tests/unit/threaded_deque_test.c]: This doesn't look like a result of `PushRight` to me.

File: libutils/threaded_deque.c (hunk 3 above)
Comment (karlhto) [reply libutils/threaded_deque.c]: Heh, some of this is directly from `s/queue/deque/g`.

**Comment by karlhto** [comment]
Thanks for thorough reviews! :)

File: tests/unit/threaded_deque_test.c (hunk 6 above)
Comment (karlhto) [reply tests/unit/threaded_deque_test.c]: Whoops. Thanks for noticing!

File: libutils/threaded_deque.c (hunk 4 above)
Comment (karlhto) [reply libutils/threaded_deque.c]: This case only happens when the right index is lower than the left index, in which case we need to make sure we have a continuous block of memory.

`memmove` takes parameters `dest`, `src`, `size`, in that order.

The first position is the position it will be moved to, which is at the first position of the newly allocated memory, at offset `old_capacity` in `deque->data`.

File: libutils/threaded_deque.c [hunk 7]
```diff
@@ -0,0 +1,641 @@
...
+        free(deque);
+    }
+}
+
+bool ThreadedDequePopLeft(ThreadedDeque *deque,
+                          void **item,
+                          int timeout)
+{
+    assert(deque != NULL);
+
```
[the remark this answers is not in the document]
Comment (karlhto) [reply libutils/threaded_deque.c]: Wasn't there a discussion about this is the `ThreadedQueue` PR? I thought we decided that the data structures should be able to hold `NULL` items. I also have a test for pushing `NULL` that actually pushes it as a result of that discussion.

File: libutils/threaded_deque.c (hunk 2 above)
Comment (karlhto) [reply libutils/threaded_deque.c]: I assume you mean that the function should be "flatter" as a result, right?

File: tests/unit/threaded_deque_test.c (hunk 6 above)
Comment (karlhto) [reply tests/unit/threaded_deque_test.c]: Wait, yes it is! You push to the right index.

Comment (karlhto) [reply tests/unit/threaded_deque_test.c]: It looks confusing, I know.

File: libutils/threaded_deque.c [hunk 8]
```diff
@@ -0,0 +1,641 @@
...
+    return deque;
+}
+
+void ThreadedDequeDestroy(ThreadedDeque *deque)
+{
+    if (deque != NULL)
+    {
+        ThreadLock(deque->lock);
+        DestroyRange(deque, deque->left, deque->right);
+        ThreadUnlock(deque->lock);
```
[the remark this answers is not in the document]
Comment (karlhto) [reply libutils/threaded_deque.c]: I agree, and we talked about this in the previous PR as well.

This function should only be called when you are sure all the threads are joined and the program is about to die.

**Comment by karlhto** [comment]
@olehermanse @vpodzime I have addressed you concerns.

@cf-bottom jenkins with exotics

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2183/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/2183/

File: libutils/threaded_deque.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (vpodzime) [reply libutils/threaded_deque.c]: Yeah, we discussed this in the `Queue` case. There's no good way to make these things thread-safe without a lock that would be outside of the `Queue/Dequeue` structure which is something we want to avoid.

File: libutils/threaded_deque.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (vpodzime) [reply libutils/threaded_deque.c]: There's a difference between `item` and `*item` though. ;-)

File: libutils/threaded_deque.c (hunk 2 above)
Comment (vpodzime) [reply libutils/threaded_deque.c]: Exactly.

File: libutils/threaded_deque.c [hunk 9]
```diff
@@ -0,0 +1,643 @@
...
+        return true;
+    }
+
+    int ret = true;
+
+    if (timeout == 0)
+    {
+        ret = false;
+    }
```
Comment (vpodzime) [inline libutils/threaded_deque.c]: This can also just unlock and return `false`.

File: libutils/threaded_deque.c (hunk 4 above)
Comment (vpodzime) [reply libutils/threaded_deque.c]: I see. So it's moving the "overflow" items that are at the beginning of the data segment, but are actually at the end (right side) of the queue in the "abstract picture". Could you please add an explanatory comment above this call?

File: tests/unit/threaded_deque_test.c (hunk 6 above)
Comment (vpodzime) [reply tests/unit/threaded_deque_test.c]: So you push to the right and the item is on the left? :confused:

Comment (karlhto) [reply tests/unit/threaded_deque_test.c]: It pushes to the _right end_ of the queue. This means that since the array has wrapped around by first pushing to the left, it looks as if you are pushing on the left, I guess?

If you had only do pushes right, it would look like this:
```
|><--------------|
|>x<-------------|
```
instead of:
```
|<-------------->|
|x<------------->|
```

**Comment by karlhto** [comment]
@cf-bottom jenkins with exotics

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2187/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/2187/

File: tests/unit/threaded_deque_test.c (hunk 6 above)
Comment (vpodzime) [reply tests/unit/threaded_deque_test.c]: OK, analyzed this with @olehermanse and it makes sense. Maybe the legend should explicitly say that `<` and `>` are positions where the `Pop()` would take its item from not where `Push()` would push its item. That would also make it clear that `<` is actually also `x` (occupied memory).

File: tests/unit/threaded_deque_test.c [hunk 10]
```diff
@@ -0,0 +1,426 @@
...
+ *      x : used memory                 *
+ *      - : unused memory               */
+
+static void test_push_pop(void)
+{
+    // Initialised with DEFAULT_CAPACITY = 16
+    ThreadedDeque *deque = ThreadedDequeNew(0, free);
+    // |^---------------|
+    ThreadedDequePushLeft(deque, xstrdup("1"));
+    // |<-------------->|
```
Comment (olehermanse) [inline tests/unit/threaded_deque_test.c]: @karlhto I still find this thoroughly confusing. After 1 push, shouldn't `left == right`, since a PopLeft or a PopRight would both pop the same index?

Comment (karlhto) [reply tests/unit/threaded_deque_test.c]: Alright, so:

The right index is the first empty element in a continuous block of data (that can wrap around the allocated memory area). I made it this way so I wouldn't have to have a special case when `size` is `0`, where after 

I also think doing it differently will mess with the internals of the array. It makes sense to me that you should end up at different indices when popping from the left and from the right, e.g.:
```c
/* Starting position (with one element in the data structure), where:
 *     < is right index (first empty element)
 *     - is empty
 *     > is left index (first non-empty element)
 *     ^ is both indexes when array is empty
 */
|<-------------->|

// Pop left:
|^---------------|

// Pop right
|---------------^|
```

Since we already have the `size` parameter, it is not a big deal to change.

Comment (vpodzime) [reply tests/unit/threaded_deque_test.c]: I think the legend is missing the information that `<` is actually a first **empty** element while `>` is a **non-empty** element.

**Comment by vpodzime** [comment]
I'd suggest not wasting more time on the nitpicks here. @karlhto please add the empty/non-empty info to the legend and let's merge this. I am now not even sure we will need this structure for something.

**Comment by karlhto** [comment]
@cf-bottom Jenkins with exotics (actual last time)

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2256/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2256/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2256/
